File: gs/src/main/java/com/pluxity/device/GsDeviceService.java
package com.pluxity.device;

import com.pluxity.cctv.CctvService;
import com.pluxity.cctv.dto.CctvResponse;
import com.pluxity.cctv.entity.DeviceCctv;
import com.pluxity.cctv.repository.DeviceCctvRepository;
import com.pluxity.device.dto.*;
import com.pluxity.device.entity.Device;
import com.pluxity.device.entity.DeviceCategory;
import com.pluxity.device.service.DeviceCategoryService;
import com.pluxity.feature.dto.FeatureResponse;
import com.pluxity.file.dto.FileResponse;
import com.pluxity.file.service.FileService;
import com.pluxity.global.annotation.CheckPermissionCategory;
import com.pluxity.global.constant.ErrorCode;
import com.pluxity.global.exception.CustomException;
import com.pluxity.global.utils.MappingUtils;
import com.pluxity.permission.ResourceType;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
public class GsDeviceService {

    private final GsDeviceRepository repository;
    private final DeviceCategoryService deviceCategoryService;
    private final DeviceCctvRepository deviceCctvRepository;
    private final CctvService cctvService;
    private final FileService fileService;

    @Transactional
    public String save(GsDeviceCreateRequest request) {
        DeviceCategory category =
                request.categoryId() != null ? deviceCategoryService.findById(request.categoryId()) : null;

        return repository.save(new GsDevice(request.id(), category, request.name())).getId();
    }

    @Transactional(readOnly = true)
    @CheckPermissionCategory(categoryResourceType = ResourceType.DEVICE_CATEGORY)
    public GsDeviceResponse findById(String id) {
        GsDevice gsDevice = getDevice(id);
        return createResponse(gsDevice, getThumbnailFile(gsDevice));
    }

    private FileResponse getThumbnailFile(Device gsDevice) {
        return Optional.ofNullable(gsDevice.getCategory())
                .map(DeviceCategory::getIconFileId)
                .map(fileService::getFileResponse)
                .orElse(null);
    }

    private GsDevice getDevice(String id) {
        return repository
                .findById(id)
                .orElseThrow(() -> new CustomException(ErrorCode.NOT_FOUND_DEVICE, id));
    }

    @Transactional(readOnly = true)
    @CheckPermissionCategory(categoryResourceType = ResourceType.DEVICE_CATEGORY)
    public List<GsDeviceResponse> findAll() {
        List<GsDevice> gsDevices = repository.findAll();
        List<DeviceCategory> categoryList =
                gsDevices.stream().map(Device::getCategory).filter(Objects::nonNull).toList();
        Map<Long, FileResponse> fileMap =
                MappingUtils.getFileMapByIds(categoryList, v -> Stream.of(v.getIconFileId()), fileService);
        return gsDevices.stream()
                .map(
                        gsDevice ->
                                createResponse(
                                        gsDevice,
                                        gsDevice.getCategory() != null
                                                ? fileMap.get(gsDevice.getCategory().getIconFileId())
                                                : null))
                .collect(Collectors.toList());
    }

    private static GsDeviceResponse createResponse(GsDevice gsDevice, FileResponse thumbnailFile) {
        return GsDeviceResponse.builder()
                .id(gsDevice.getId())
                .name(gsDevice.getName())
                .feature(gsDevice.getFeature() != null ? FeatureResponse.from(gsDevice.getFeature()) : null)
                .deviceCategory(
                        gsDevice.getCategory() != null
                                ? DeviceCategoryResponseWithoutChildren.from(gsDevice.getCategory(), thumbnailFile)
                                : null)
                .build();
    }

    @Transactional
    public void update(String id, GsDeviceUpdateRequest request) {
        GsDevice device = getDevice(id);
        device.update(request.name());

        if (request.categoryId() != null) {
            DeviceCategory deviceCategory = deviceCategoryService.findById(request.categoryId());
            device.changeCategory(deviceCategory);
        }
    }

    @Transactional
    public void putUpdate(String id, GsDeviceUpdateRequest request) {
        GsDevice device = getDevice(id);
        device.putUpdate(request.name());
        DeviceCategory category =
                MappingUtils.findByIdIfExists(request.categoryId(), deviceCategoryService::findById);
        device.changeCategory(category);
    }

    @Transactional
    public void delete(String id) {
        GsDevice device = getDevice(id);
        device.clearAllRelations();
        deviceCctvRepository.deleteByDevice(device);
        repository.delete(device);
    }

    @Transactional
    public void assignCategory(String deviceId, Long categoryId) {
        GsDevice device = getDevice(deviceId);
        DeviceCategory deviceCategory = deviceCategoryService.findById(categoryId);

        device.changeCategory(deviceCategory);
        log.info("디바이스 [{}}에 카테고리 [{}]가 할당되었습니다.", deviceId, categoryId);
    }

    @Transactional
    public void removeCategory(String deviceId) {
        GsDevice device = getDevice(deviceId);

        if (device.getCategory() == null) {
            throw new CustomException(ErrorCode.NOT_FOUND_DEVICE_CATEGORY, deviceId);
        }

        device.changeCategory(null);
        log.info("디바이스 [{}]에서 카테고리가 제거되었습니다.", deviceId);
    }

    @Transactional(readOnly = true)
    public List<CctvResponse> getCctvByDeviceId(String deviceId) {
        GsDevice device = getDevice(deviceId);
        List<DeviceCctv> deviceCctvs = deviceCctvRepository.findByDevice(device);
        return deviceCctvs.stream()
                .map(DeviceCctv::getCctv)
                .map(
                        cctv ->
                                CctvResponse.builder()
                                        .id(cctv.getId())
                                        .name(cctv.getName())
                                        .url(cctv.getUrl())
                                        .feature(
                                                cctv.getFeature() != null ? FeatureResponse.from(cctv.getFeature()) : null)
                                        .deviceCategory(
                                                cctv.getCategory() != null
                                                        ? DeviceCategoryResponseWithoutChildren.from(
                                                                cctv.getCategory(), getThumbnailFile(cctv))
                                                        : null)
                                        .build())
                .toList();
    }

    @Transactional
    public void assignCctvToDevice(String deviceId, GsDeviceCctvUpdateRequest request) {
        GsDevice device = getDevice(deviceId);
        List<String> existIds =
                deviceCctvRepository.findByDevice(device).stream().map(v -> v.getCctv().getId()).toList();

        List<String> requestIds = new ArrayList<>(request.cctvIds());

        // 추가할 cctv id
        List<DeviceCctv> saveList =
                requestIds.stream()
                        .filter(id -> !existIds.contains(id))
                        .map(
                                cctvId ->
                                        DeviceCctv.builder().cctv(cctvService.findById(cctvId)).device(device).build())
                        .toList();

        // 삭제할 cctv id
        List<String> removeList = existIds.stream().filter(id -> !requestIds.contains(id)).toList();

        // 추가
        if (!saveList.isEmpty()) {
            deviceCctvRepository.saveAll(saveList);
        }

        // 삭제
        if (!removeList.isEmpty()) {
            deviceCctvRepository.deleteByCctvIdIn(removeList);
        }
    }
}

File: gs/src/main/kotlin/com/pluxity/device/GsDeviceService.kt
package com.pluxity.device

import com.pluxity.cctv.CctvService
import com.pluxity.cctv.dto.CctvResponse
import com.pluxity.cctv.dto.toCctvResponse
import com.pluxity.cctv.entity.DeviceCctv
import com.pluxity.cctv.repository.DeviceCctvRepository
import com.pluxity.device.dto.GsDeviceCctvUpdateRequest
import com.pluxity.device.dto.GsDeviceCreateRequest
import com.pluxity.device.dto.GsDeviceResponse
import com.pluxity.device.dto.GsDeviceUpdateRequest
import com.pluxity.device.dto.toGsDeviceResponse
import com.pluxity.device.entity.Device
import com.pluxity.device.entity.DeviceCategory
import com.pluxity.device.service.DeviceCategoryService
import com.pluxity.file.dto.FileResponse
import com.pluxity.file.service.FileService
import com.pluxity.global.annotation.CheckPermissionCategory
import com.pluxity.global.constant.ErrorCode
import com.pluxity.global.exception.CustomException
import com.pluxity.global.utils.MappingUtils
import com.pluxity.permission.ResourceType
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.stream.Stream

private val log = KotlinLogging.logger {}

@Service
class GsDeviceService(
    private val repository: GsDeviceRepository,
    private val deviceCategoryService: DeviceCategoryService,
    private val deviceCctvRepository: DeviceCctvRepository,
    private val cctvService: CctvService,
    private val fileService: FileService,
) {
    @Transactional
    fun save(request: GsDeviceCreateRequest): String {
        val category = request.categoryId?.let { deviceCategoryService.findById(request.categoryId) }
        return repository.save(GsDevice(request.id, category, request.name)).id
    }

    @Transactional(readOnly = true)
    @CheckPermissionCategory(categoryResourceType = ResourceType.DEVICE_CATEGORY)
    fun findById(id: String): GsDeviceResponse = getDevice(id).toGsDeviceResponse(getThumbnailFile(getDevice(id)))

    private fun getThumbnailFile(gsDevice: Device): FileResponse? =
        gsDevice.category?.let {
            fileService.getFileResponse(it.iconFileId)
        }

    private fun getDevice(id: String): GsDevice =
        repository.findByIdOrNull(id)
            ?: throw CustomException(ErrorCode.NOT_FOUND_DEVICE, id)

    @Transactional(readOnly = true)
    @CheckPermissionCategory(categoryResourceType = ResourceType.DEVICE_CATEGORY)
    fun findAll(): List<GsDeviceResponse> {
        val gsDevices = repository.findAll()
        val categoryList =
            gsDevices
                .mapNotNull { it.category }
        val fileMap =
            MappingUtils.getFileMapByIds(
                categoryList,
                { v: DeviceCategory -> Stream.of(v.iconFileId) },
                fileService,
            )
        return gsDevices.map {
            it.toGsDeviceResponse(fileMap[it.category?.iconFileId])
        }
    }

    @Transactional
    fun update(
        id: String,
        request: GsDeviceUpdateRequest,
    ) {
        val device = getDevice(id)
        device.update(request.name)
        request.categoryId?.let { device.changeCategory(deviceCategoryService.findById(it)) }
    }

    @Transactional
    fun putUpdate(
        id: String,
        request: GsDeviceUpdateRequest,
    ) {
        val device = getDevice(id)
        device.putUpdate(request.name)
        val category = request.categoryId?.let { deviceCategoryService.findById(it) }
        device.changeCategory(category)
    }

    @Transactional
    fun delete(id: String) {
        val device = getDevice(id)
        device.clearAllRelations()
        deviceCctvRepository.deleteByDevice(device)
        repository.deleteById(device.id)
    }

    @Transactional
    fun assignCategory(
        deviceId: String,
        categoryId: Long,
    ) {
        val device = getDevice(deviceId)
        device.changeCategory(deviceCategoryService.findById(categoryId))
        log.info { "디바이스 [$deviceId]에 카테고리 [$categoryId]가 할당되었습니다." }
    }

    @Transactional
    fun removeCategory(deviceId: String) {
        val device = getDevice(deviceId)
        device.category ?: throw CustomException(ErrorCode.NOT_FOUND_ASSIGN_DEVICE_CATEGORY, deviceId)
        device.changeCategory(null)
        log.info { "디바이스 [$deviceId]에서 카테고리가 제거되었습니다." }
    }

    @Transactional(readOnly = true)
    fun getCctvByDeviceId(deviceId: String): List<CctvResponse> {
        val device = getDevice(deviceId)
        val deviceCctvs = deviceCctvRepository.findByDevice(device)
        return deviceCctvs
            .map { it.cctv }
            .map { it.toCctvResponse(getThumbnailFile(it)) }
    }

    @Transactional
    fun assignCctvToDevice(
        deviceId: String,
        request: GsDeviceCctvUpdateRequest,
    ) {
        val device = getDevice(deviceId)
        val existIds = deviceCctvRepository.findByDevice(device).map { it.cctv.id!! }

        // 추가할 cctv id
        val saveList =
            request.cctvIds
                .filter { !existIds.contains(it) }
                .map { DeviceCctv(cctv = cctvService.findById(it), device = device) }

        // 삭제할 cctv id
        val removeList: List<String> = existIds.filter { !request.cctvIds.contains(it) }

        // 추가
        if (saveList.isNotEmpty()) {
            log.info { "${deviceId}에 추가할 cctvId $saveList" }
            deviceCctvRepository.saveAll(saveList)
        }

        // 삭제
        if (removeList.isNotEmpty()) {
            log.info { "${deviceId}에서 제거할 cctvId $removeList" }
            deviceCctvRepository.deleteByCctvIdIn(removeList)
        }
    }
}
